sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -850,7 +850,7 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
     val errorMsg1 =
       s"""
          |Input to function array_contains should have been array followed by a
-         |value with same element type, but it's [array<int>, decimal(29,29)].
+         |value with same element type, but it's [array<int>, decimal(38,29)].
```

Comment:
why precision becomes 38 in this case?

Reply:
https://github.com/apache/spark/blob/1fc353d51a62cb554e6af23dbc9a613e214e3af1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L864-L869
For query `array_contains(array(1), .01234567890123456790123456780)`\
`e.inputTypes` will return `Seq(Array(Decimal(38,29)), Decimal(38,29))` and above code will cast `.01234567890123456790123456780` as `Decimal(38,29)`. 
Previously, when we were using `findWiderTypeForTwo`, decimal types were not getting upcasted but `findWiderTypeWithoutStringPromotionForTwo` will successfully upcast DecimalType

Reply:
cc @cloud-fan

Reply:
> Previously, when we were using findWiderTypeForTwo

Before this PR, we were using `findTightestCommonType`. Why do we add cast but still can't resolve `ArrayContains`?

Reply:
Do you mean why in above test case query, `ArrayContains` is throwing `AnalysisException` instead of casting integer to Decimal?

An integer cannot be casted to decimal with scale > 28.

```
decimalWith28Zeroes = 1.0000000000000000000000000000
SELECT array_contains(array(1), decimalWith28Zeroes);
Result =>> true
```

```
decimalWith29Zeroes = 1.00000000000000000000000000000
SELECT array_contains(array(1), decimalWith29Zeroes);
Result =>> AnalysisException
```

Reply:
cc @maropu @cloud-fan

Reply:
Yea I get that we can't do cast here. My question is: since we can't do cast, we should leave the expression un-touched. But now we add cast to one side and leave the expression unresolved. Where do we add that uesless cast?

Reply:
https://github.com/apache/spark/blob/1fc353d51a62cb554e6af23dbc9a613e214e3af1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L864-L869

This code is to cast left and right expression one by one. Here,
- `e.childern` is `Seq( array<int>, decimal(29,29))`, and
- `e.inputTypes` will return `Seq(array<decimal(38,29)>, decimal(38,29))`

`impicitCast(array<int>, array<decimal(38,29)>)` will return `None`, since `int` can't be casted to `decimal(38,29)`.

Reply:
Above code is creating new expression by updating only right child.

Reply:
ah thanks for finding this out!